modules/openapi-generator/src/main/java/org/openapitools/codegen/utils/ModelUtils.java

```diff
@@ -1080,12 +1080,55 @@ public static Schema unaliasSchema(OpenAPI openAPI,
         return schema;
     }
 
-    public static Schema getAdditionalProperties(Schema schema) {
-        if (schema.getAdditionalProperties() instanceof Schema) {
-            return (Schema) schema.getAdditionalProperties();
+    /**
+     * Returns the additionalProperties Schema for the specified input schema.
+     * 
+     * The additionalProperties keyword is used to control the handling of additional, undeclared
+     * properties, that is, properties whose names are not listed in the properties keyword.
+     * The additionalProperties keyword may be either a boolean or an object.
+     * If additionalProperties is a boolean and set to false, no additional properties are allowed.
+     * By default when the additionalProperties keyword is not specified in the input schema,
+     * any additional properties are allowed. This is equivalent to setting additionalProperties
+     * to the boolean value True or setting additionalProperties: {}
+     * 
+     * @param openAPI the object that encapsulates the OAS document.
+     * @param schema the input schema that may or may not have the additionalProperties keyword.
+     * @return the Schema of the additionalProperties. The null value is returned if no additional
+     *         properties are allowed.
+     */
+    public static Schema getAdditionalProperties(OpenAPI openAPI, Schema schema) {
+        Object addProps = schema.getAdditionalProperties();
+        if (addProps instanceof Schema) {
+            return (Schema) addProps;
+        }
+        if (addProps == null) {
+            Map<String, Object> extensions = openAPI.getExtensions();
+            if (extensions != null) {
+                // Get original swagger version from OAS extension.
+                // Note openAPI.getOpenapi() is always set to 3.x even when the document
+                // is converted from a OAS/Swagger 2.0 document.
+                // https://github.com/swagger-api/swagger-parser/pull/1374
+                Object ext = extensions.get("x-original-swagger-version");
+                if (ext instanceof String) {
+                    SemVer version = new SemVer((String)ext);
+                    if (version.major == 2) {
+                        // The OAS version 2 parser sets Schema.additionalProperties to the null value
+                        // even if the OAS document has additionalProperties: true|false
+                        // So we are unable to determine if additional properties are allowed or not.
```

Comment:
Should we handle this in a preprocessor before a v2 spec is converted to v3?

Reply:
This is a fine fix for our current use case.
It looks like you are working on a 2.0 fix [here](https://github.com/swagger-api/swagger-parser/pull/1370)